### ISSUE TITLE
Existing upmaps are lost

### ISSUE THREAD

**Issue by patrakov** [laimis9133/plankton-swarm#1]
Suppose that there is a PG with an existing upmap, and it is on an osd with too-high utilization. Then, the script would do this:

```
      echo "ceph osd pg-upmap-items $pg $osd $selected_osd"
      echo "ceph osd pg-upmap-items $pg $osd $selected_osd" >> "$balance_file"
```

However, this loses the existing upmaps that this PG might have for other replicas or shards, and therefore, this code can cause unwanted data movement, including the one that undoes earlier optimizations.

**Comment by laimis9133** [comment]
Hey, thanks for cheking out the script! If there is a PG with an existing upmap it filtered out as only active+clean PGs are picked up:
`pgs=$(ceph pg dump | grep ",$osd]" | grep -P 'active\+clean(?!\+)' | awk '{print $1, $19}' | head -n "$((pg_limit * pg_pre_fetch))" | shuf)`

If my understanding is correct a pg with an existing upmap will be in remapped state and not pass through `grep -P 'active\+clean(?!\+)'`?

**Comment by patrakov** [comment]
No, the remapped state is only there when the data is not at its final destination. If you add an upmap, it will cause the data to move, and after that, the PG will be active+clean, but the data will not be where the CRUSH map wants it to be.

**Comment by laimis9133** [comment]
Hey, coming back to this - I've just commited a rewamp of the script. 
Amongst other changes I have added options to respect existing pg upmaps. Two available flags now that you can pass to the script: `check-upmaps` and `keep-upmaps`
`keep-upmaps` will simply skip any pgs that already have upmaps and is the safest option.
`check-upmaps` will gather a list of OSDs in existing upmaps and skip their parent nodes as targets. This will avoid a redo of any (already done) data optimisations.

Both flags are set to `false` as default because the whole idea of plankton is to have a simple manual balancer where you want to move data from OSD to OSD overriding any existing upmaps or CRUSH rules or reweights (becasue you know better how things should be).

One sidenode to bring some background: we generate regular (cron) small data movements via plankton swarm with Ceph balancer switched off (19.2.0 issues) and also found it useful for larger batch movements when performing node/disk replacements. 
Its is more like a tool for targeted, one-time tasks—a shovel for specific digs—rather than a background system.
